From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Liangliang Sui <coolsui.coding@gmail.com>
Date: Tue, 19 Mar 2024 21:04:57 +0800
Subject: [PATCH 094/226] Asynchronous initialization of shader cache.

Put the shader initialization logic into a separate thread and lock it
to reduce the load on RenderThread. The interval between initialization
and use of ShaderCache is a long time (~50ms). This period can be enough
for ShaderCache to execute the initialization logic.

Due to the protection of mMutex, there will be no thread safety issues.

Change-Id: I7fb3ab06d0003eeba1a542d872177ab80e7d4d73
Signed-off-by: Liangliang Sui <coolsui.coding@gmail.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 libs/hwui/pipeline/skia/ShaderCache.cpp | 29 +++++++++++++++----------
 1 file changed, 17 insertions(+), 12 deletions(-)

diff --git a/libs/hwui/pipeline/skia/ShaderCache.cpp b/libs/hwui/pipeline/skia/ShaderCache.cpp
index 22f59a67bc..9edd42fd78 100644
--- a/libs/hwui/pipeline/skia/ShaderCache.cpp
+++ b/libs/hwui/pipeline/skia/ShaderCache.cpp
@@ -35,10 +35,10 @@ namespace uirenderer {
 namespace skiapipeline {
 
 // Cache size limits.
-static const size_t maxKeySize = 1024;
-static const size_t maxValueSize = 2 * 1024 * 1024;
-static const size_t maxTotalSize = 4 * 1024 * 1024;
-static_assert(maxKeySize + maxValueSize < maxTotalSize);
+static const size_t kMaxKeySize = 1024;
+static const size_t kMaxValueSize = 2 * 1024 * 1024;
+static const size_t kMaxTotalSize = 4 * 1024 * 1024;
+static_assert(kMaxKeySize + kMaxValueSize < kMaxTotalSize);
 
 ShaderCache::ShaderCache() {
     // There is an "incomplete FileBlobCache type" compilation error, if ctor is moved to header.
@@ -89,12 +89,17 @@ void ShaderCache::initShaderDiskCache(const void* identity, ssize_t size) {
     // or snapshot migration. Also, program binaries may not work well on some
     // desktop / laptop GPUs. Thus, disable the shader disk cache for emulator builds.
     if (!Properties::runningInEmulator && mFilename.length() > 0) {
-        mBlobCache.reset(new FileBlobCache(maxKeySize, maxValueSize, maxTotalSize, mFilename));
-        validateCache(identity, size);
-        mInitialized = true;
-        if (identity != nullptr && size > 0 && mIDHash.size()) {
-            set(&sIDKey, sizeof(sIDKey), mIDHash.data(), mIDHash.size());
-        }
+        std::thread deferredInitThread([this, identity, size]() {
+            std::lock_guard lock(mMutex);
+            mBlobCache.reset(
+                new FileBlobCache(kMaxKeySize, kMaxValueSize, kMaxTotalSize, mFilename));
+            validateCache(identity, size);
+            mInitialized = true;
+            if (identity != nullptr && size > 0 && mIDHash.size()) {
+                set(&sIDKey, sizeof(sIDKey), mIDHash.data(), mIDHash.size());
+            }
+        });
+        deferredInitThread.detach();
     }
 }
 
@@ -120,7 +125,7 @@ sk_sp<SkData> ShaderCache::load(const SkData& key) {
     size_t valueSize = mBlobCache->get(key.data(), keySize, valueBuffer, mObservedBlobValueSize);
     int maxTries = 3;
     while (valueSize > mObservedBlobValueSize && maxTries > 0) {
-        mObservedBlobValueSize = std::min(valueSize, maxValueSize);
+        mObservedBlobValueSize = std::min(valueSize, kMaxValueSize);
         void* newValueBuffer = realloc(valueBuffer, mObservedBlobValueSize);
         if (!newValueBuffer) {
             free(valueBuffer);
@@ -199,7 +204,7 @@ void ShaderCache::store(const SkData& key, const SkData& data, const SkString& /
 
     size_t valueSize = data.size();
     size_t keySize = key.size();
-    if (keySize == 0 || valueSize == 0 || valueSize >= maxValueSize) {
+    if (keySize == 0 || valueSize == 0 || valueSize >= kMaxValueSize) {
         ALOGW("ShaderCache::store: sizes %d %d not allowed", (int)keySize, (int)valueSize);
         return;
     }
